### PR TITLE
(CFACT-140) Add acceptance tests for Facter executable

### PR DESCRIPTION
**Pull request by whopper**
This commit adds a new suite of tests around the Facter executable.
Specifically, new tests are added for each of Facter's command-line
options.

### CODE REVIEW

**Comment by whopper** [comment]
I'm putting this up for initial review only, since the new facter CI pipeline isn't quite ready yet. These tests pass with the invocation: `bundle exec rake ci:test:packages SHA=a18dc6a240cf1c89f7fa8ce74f7e15652ac26558 TESTS=./tests/options CONFIG=./config/nodes/trusty.yaml` where the SHA is one of the latest builds from builds.puppetlabs.lan/cfacter. Once the name is completely changed over to Facter, it will draw packages from builds.puppetlabs.lan/facter instead.

I haven't yet written a test for the `--log-level` option. Any advice on how best to test `warn`, `error` and `fatal` args?

File: acceptance/tests/options/trace.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+  teardown do
+    on(agent, "rm -f '#{custom_fact}'")
+  end
+
+  step "--trace option should provide a backtrace for a custom fact with errors"
+  begin
+    on(agent, "FACTERLIB=#{custom_dir} cfacter --trace custom_fact")
+  rescue Exception => e
+    assert_match(/backtrace:\n\t#{custom_fact}/, e.message, "Expected a backtrace for erroneous custom fact")
```
Comment (whopper) [inline acceptance/tests/options/trace.rb]: Is this an insane way to test this?

File: acceptance/tests/options/trace.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+  teardown do
+    on(agent, "rm -f '#{custom_fact}'")
+  end
+
+  step "--trace option should provide a backtrace for a custom fact with errors"
+  begin
+    on(agent, "FACTERLIB=#{custom_dir} cfacter --trace custom_fact")
+  rescue Exception => e
+    assert_match(/backtrace:\s+#{custom_fact}/, e.message, "Expected a backtrace for erroneous custom fact")
```
Comment (whopper) [inline acceptance/tests/options/trace.rb]: Is this an insane way to test this?

**Comment by MikaelSmith** [comment]
I'd limit testing --log-level to verifying that the messages are accepted.

**Comment by whopper** [comment]
@MikaelSmith the second commit I pushed just now is an experiment to extract the external / custom fact pathing logic away from the tests themselves, since there was a lot of duplication. Do you think that's necessary, or the right way to go about it?

**Comment by MikaelSmith** [comment]
Sounds useful.

**Comment by whopper** [comment]
Another note: the test that checks the --no-custom-fact option is legitimately failing atm.

**Comment by MikaelSmith** [comment]
Is the --no-custom-fact option trivial to fix? I'd rather not have an acceptance test failing for any extended period, so we could either pull that test out to go with the other ticket, or prepare the fix as part of this PR.

**Comment by peterhuene** [comment]
It is trivial.  facter.cc:240: `if (ruby)` => `if (ruby && !vm.count("no-external-facts"))`

**Comment by whopper** [comment]
@MikaelSmith @peterhuene I'll put a PR up for that bit first, and then s/cfacter/facter/ for all of these tests.

**Comment by whopper** [comment]
PR for the --no-custom-fact fix is here: https://github.com/puppetlabs/facter/pull/883

**Comment by peterhuene** [comment]
:+1: from me.  I'll let Michael also review before merging.

File: acceptance/lib/facter/acceptance/user_fact_utils.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            'C:/ProgramData/PuppetLabs/facter/custom'
+          end
+        else
+          '/opt/puppetlabs/facter/custom'
+        end
+      end
+
+      # Retreive the path of the standard facts.d directory.
+      #
+      def get_factsd_dir(platform, version)
```
Comment (MikaelSmith) [inline acceptance/lib/facter/acceptance/user_fact_utils.rb]: Could you update runs_external_facts_once to use this?

**Comment by whopper** [comment]
@MikaelSmith sure thing, updated.

**Comment by MikaelSmith** [comment]
Only acceptance test changes, so merging.

**Pull request merged**